dafny-lang/dafny
# Let the IDE translate to and resolve Boogie only when attempting to verify

### Changes

- Let the IDE translate to and resolve Boogie only when attempting to verify, instead of doing those steps for all code in the project right after resolution, and before allowing any verification to be done.

  - Using the experimental verification caching feature to detect which tasks were already verified before the user triggers any verification, no longer works. This is because that caching relies on pre-verification which we now only do when the user triggers verification. However, when trying to verify these cached tasks, they'll be completed immediately after pre-verification finishes. In the future, we will have to implement caching of verification based on the Dafny AST if we want it to show up before pre-verification.
  - The Dafny AST nodes must now specify whether they can be verified, even before being translated to Boogie. This creates a slight duplication of effort, since this information was implicitly also provided by the Boogie translation
- Properly close files after they are read, both `.dfy` and `.dfyconfig.toml` files
- XUnit tests now all run sequentially. This does not seem to have a significant affect on total test runtime and makes debugging test timeouts easier because it excludes lack of available threads as a cause.
- The class `ConstantField` is moved to a separate file
- Add AutoGeneratedToken where they were missing.
- Configure the `dotnet test` call on the CLI so individual server tests won't run for more than 6 minutes.
- Split the existing `OnChange` mode into `OnChangeFile` and `OnChangeProject`, which let verification automatically run if either their containing file or containing project has changed, respectively.
- `Stringify`, a method to help with debugging, has been improved
- Small cleanup of `CounterExampleTest`, changing a `List<Action>` into an `Action`
- Many improvements to logging

### Testing
- Existing tests have been adapted for the new behavior
- The performance part improvement of this PR is not tested, since we don't have a good framework for testing performance.
- Add tests for the behavior of `OnChangeFile` and `OnChangeProject` mode.
- Add tests for `Stringify`
- Add tests for `FindNode`

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>
